src/utils/batch-operate/SessionQueue.ts

```diff
@@ -0,0 +1,153 @@
+		client: AirtopClient;
+		sessionConfig?: AirtopSessionConfigV1;
+		onError?: (error: Error | string) => void;
+	}) {
+		if (
+			!Number.isInteger(maxConcurrentSessions) ||
+			maxConcurrentSessions <= 0
+		) {
+			throw new Error("maxConcurrentSessions must be a positive integer");
+		}
+
+		this.maxConcurrentSessions = maxConcurrentSessions;
+		this.runEmitter = runEmitter;
+		this.maxWindowsPerSession = maxWindowsPerSession;
+		this.sessionConfig = sessionConfig;
+		this.initialBatches = initialBatches;
+		this.operation = operation;
+		this.onError = onError;
+		this.processingPromise = null;
+
+		this.client = client;
+	}
+
+	public async addUrlsToBatchQueue(
+		newBatch: BatchOperationUrl[],
+	): Promise<void> {
+		// Distribute new URLs across batches
+		const newBatches = distributeUrlsToBatches(
+			newBatch,
+			this.maxConcurrentSessions,
+		);
+
+		this.client.log(
+			`Adding new batches to queue: ${JSON.stringify(newBatches)}`,
+		);
+		this.batchQueue.push(...newBatches);
+
+		// Update existing processing promise
+		this.processingPromise = this.processPendingBatches();
+	}
+
+	public async processInitialBatches(): Promise<void> {
+		this.batchQueue = [...this.initialBatches];
+		this.processingPromise = this.processPendingBatches();
+		await this.processingPromise;
+	}
+
+	public async waitForProcessingToComplete(): Promise<void> {
+		await this.processingPromise;
+	}
+
+	private async processPendingBatches(): Promise<void> {
+		while (this.batchQueue.length > 0) {
+			if (this.activePromises.length < this.maxConcurrentSessions) {
+				const batch = this.batchQueue.shift();
+				if (!batch || batch.length === 0) break;
+
+				const promise = (async () => {
+					let sessionId: string | undefined;
+					try {
+						const { data: session } = await this.client.sessions.create({
+							configuration: this.sessionConfig,
+						});
+						sessionId = session.id;
+
+						const queue = new WindowQueue(
+							this.maxWindowsPerSession,
+							this.runEmitter,
+							sessionId,
+							this.client,
+							this.operation,
+							this.onError,
+						);
+						await queue.processInBatches(batch);
+					} catch (error) {
+						if (this.onError) {
+							this.onError(
+								error instanceof Error ? error.message : String(error),
+							);
+						} else {
+							this.client.error(
+								error instanceof Error ? error.message : String(error),
+							);
+						}
+					} finally {
+						if (sessionId) {
+							await this.client.sessions.terminate(sessionId);
+						}
+					}
+				})();
+
+				this.activePromises.push(promise);
+
+				// Remove the promise when it completes
+				promise.finally(() => {
+					const index = this.activePromises.indexOf(promise);
+					if (index > -1) {
+						this.activePromises.splice(index, 1);
+					}
+				});
+			} else {
+				// Wait for any session to complete before starting a new one
+				await Promise.race(this.activePromises);
```

Comment:
nit. For easier readibility of the code, I will move this `else` block at the beginning of the loop and have a break instead of an if/else.
That way you can reduce the indentation levels of this loop.

Another option will be to move the big logic portion of the loop to it's own method.

This is just a personal opinion, it's perfectly fine the way it is.

The same comment applies on the windowQueue.